Panic on startup

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1091680]

goroutine 171 [running]:
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(Service).initializeChainInf
o(0xc000a4b500, 0x186ad00, 0xc000044040, 0x0, 0x0)
beacon-chain/blockchain/service.go:316 +0x2d0
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(Service).Start(0xc000a4b500
)
beacon-chain/blockchain/service.go:112 +0x1a8
created by github.com/prysmaticlabs/prysm/shared.(*ServiceRegistry).StartAll
shared/service_registry.go:44 +0x23e
```

Offending line of code: https://github.com/prysmaticlabs/prysm/blob/be5451abef69a9b501945f96e36f03b6bb8874ae/beacon-chain/blockchain/service.go#L316

Not sure how to reproduce yet, asked user for copy of beaconchain.db.

--- a/beacon-chain/blockchain/forkchoice/process_block.go
+++ b/beacon-chain/blockchain/forkchoice/process_block.go
@@ -384,7 +384,7 @@ func (s *Store) rmStatesBySlots(ctx context.Context, startSlot uint64, endSlot u
 
 	// Do not remove genesis state and epoch boundary state.
 	if startSlot%params.BeaconConfig().SlotsPerEpoch == 0 {
-		startSlot += 1
+		startSlot++
 	}
 
 	filter := filters.NewFilter().SetStartSlot(startSlot).SetEndSlot(endSlot)

--- a/beacon-chain/blockchain/forkchoice/process_block.go
+++ b/beacon-chain/blockchain/forkchoice/process_block.go
@@ -382,6 +382,11 @@ func (s *Store) rmStatesBySlots(ctx context.Context, startSlot uint64, endSlot u
 	ctx, span := trace.StartSpan(ctx, "forkchoice.rmStatesBySlots")
 	defer span.End()
 
+	// Do not remove genesis state and epoch boundary state.
+	if startSlot%params.BeaconConfig().SlotsPerEpoch == 0 {
+		startSlot++
+	}
+
 	filter := filters.NewFilter().SetStartSlot(startSlot).SetEndSlot(endSlot)
 	roots, err := s.db.BlockRoots(ctx, filter)
 	if err != nil {

--- a/beacon-chain/blockchain/forkchoice/process_block.go
+++ b/beacon-chain/blockchain/forkchoice/process_block.go
@@ -107,7 +107,7 @@ func (s *Store) OnBlock(ctx context.Context, b *ethpb.BeaconBlock) error {
 
 		startSlot := helpers.StartSlot(s.finalizedCheckpt.Epoch + 1)
 		endSlot := helpers.StartSlot(postState.FinalizedCheckpoint.Epoch+1) - 1 // Inclusive
-		if err := s.rmStatesBySlots(ctx, startSlot, endSlot); err != nil {
+		if err := s.rmStatesSinceLastFinalized(ctx, startSlot, endSlot); err != nil {
 			return errors.Wrapf(err, "could not delete states prior to finalized check point, range: %d, %d",
 				startSlot, endSlot+params.BeaconConfig().SlotsPerEpoch)
 		}
@@ -187,7 +187,7 @@ func (s *Store) OnBlockNoVerifyStateTransition(ctx context.Context, b *ethpb.Bea
 		helpers.ClearAllCaches()
 		startSlot := helpers.StartSlot(s.finalizedCheckpt.Epoch + 1)
 		endSlot := helpers.StartSlot(postState.FinalizedCheckpoint.Epoch+1) - 1 // Inclusive
-		if err := s.rmStatesBySlots(ctx, startSlot, endSlot); err != nil {
+		if err := s.rmStatesSinceLastFinalized(ctx, startSlot, endSlot); err != nil {
 			return errors.Wrapf(err, "could not delete states prior to finalized check point, range: %d, %d",
 				startSlot, endSlot+params.BeaconConfig().SlotsPerEpoch)
 		}
@@ -377,12 +377,12 @@ func (s *Store) clearSeenAtts() {
 	s.seenAtts = make(map[[32]byte]bool)
 }
 
-// rmStatesBySlots deletes the states in db in between the range of slots.
-func (s *Store) rmStatesBySlots(ctx context.Context, startSlot uint64, endSlot uint64) error {
+// rmStatesSinceLastFinalized deletes the states in db since last finalized check point.
+func (s *Store) rmStatesSinceLastFinalized(ctx context.Context, startSlot uint64, endSlot uint64) error {
 	ctx, span := trace.StartSpan(ctx, "forkchoice.rmStatesBySlots")
 	defer span.End()
 
-	// Do not remove genesis state and epoch boundary state.
+	// Do not remove genesis state or finalized state at epoch boundary.
 	if startSlot%params.BeaconConfig().SlotsPerEpoch == 0 {
 		startSlot++
 	}

--- a/beacon-chain/blockchain/forkchoice/process_block.go
+++ b/beacon-chain/blockchain/forkchoice/process_block.go
@@ -107,7 +107,7 @@ func (s *Store) OnBlock(ctx context.Context, b *ethpb.BeaconBlock) error {
 
 		startSlot := helpers.StartSlot(s.finalizedCheckpt.Epoch + 1)
 		endSlot := helpers.StartSlot(postState.FinalizedCheckpoint.Epoch+1) - 1 // Inclusive
-		if err := s.rmStatesSinceLastFinalized(ctx, startSlot, endSlot); err != nil {
+		if err := s.rmStatesOlderThanLastFinalized(ctx, startSlot, endSlot); err != nil {
 			return errors.Wrapf(err, "could not delete states prior to finalized check point, range: %d, %d",
 				startSlot, endSlot+params.BeaconConfig().SlotsPerEpoch)
 		}
@@ -187,7 +187,7 @@ func (s *Store) OnBlockNoVerifyStateTransition(ctx context.Context, b *ethpb.Bea
 		helpers.ClearAllCaches()
 		startSlot := helpers.StartSlot(s.finalizedCheckpt.Epoch + 1)
 		endSlot := helpers.StartSlot(postState.FinalizedCheckpoint.Epoch+1) - 1 // Inclusive
-		if err := s.rmStatesSinceLastFinalized(ctx, startSlot, endSlot); err != nil {
+		if err := s.rmStatesOlderThanLastFinalized(ctx, startSlot, endSlot); err != nil {
 			return errors.Wrapf(err, "could not delete states prior to finalized check point, range: %d, %d",
 				startSlot, endSlot+params.BeaconConfig().SlotsPerEpoch)
 		}
@@ -377,8 +377,8 @@ func (s *Store) clearSeenAtts() {
 	s.seenAtts = make(map[[32]byte]bool)
 }
 
-// rmStatesSinceLastFinalized deletes the states in db since last finalized check point.
-func (s *Store) rmStatesSinceLastFinalized(ctx context.Context, startSlot uint64, endSlot uint64) error {
+// rmStatesOlderThanLastFinalized deletes the states in db since last finalized check point.
+func (s *Store) rmStatesOlderThanLastFinalized(ctx context.Context, startSlot uint64, endSlot uint64) error {
 	ctx, span := trace.StartSpan(ctx, "forkchoice.rmStatesBySlots")
 	defer span.End()
 

--- a/beacon-chain/blockchain/forkchoice/process_block.go
+++ b/beacon-chain/blockchain/forkchoice/process_block.go
@@ -101,6 +101,10 @@ func (s *Store) OnBlock(ctx context.Context, b *ethpb.BeaconBlock) error {
 	if postState.FinalizedCheckpoint.Epoch > s.finalizedCheckpt.Epoch {
 		s.clearSeenAtts()
 		helpers.ClearAllCaches()
+		if err := s.db.SaveFinalizedCheckpoint(ctx, postState.FinalizedCheckpoint); err != nil {
+			return errors.Wrap(err, "could not save finalized checkpoint")
+		}
+
 		startSlot := helpers.StartSlot(s.finalizedCheckpt.Epoch + 1)
 		endSlot := helpers.StartSlot(postState.FinalizedCheckpoint.Epoch+1) - 1 // Inclusive
 		if err := s.rmStatesBySlots(ctx, startSlot, endSlot); err != nil {
@@ -109,9 +113,6 @@ func (s *Store) OnBlock(ctx context.Context, b *ethpb.BeaconBlock) error {
 		}
 
 		s.finalizedCheckpt = postState.FinalizedCheckpoint
-		if err := s.db.SaveFinalizedCheckpoint(ctx, postState.FinalizedCheckpoint); err != nil {
-			return errors.Wrap(err, "could not save finalized checkpoint")
-		}
 	}
 
 	// Update validator indices in database as needed.
@@ -381,9 +382,9 @@ func (s *Store) rmStatesBySlots(ctx context.Context, startSlot uint64, endSlot u
 	ctx, span := trace.StartSpan(ctx, "forkchoice.rmStatesBySlots")
 	defer span.End()
 
-	// Do not remove genesis state.
-	if startSlot == 0 {
-		startSlot = 1
+	// Do not remove genesis state and epoch boundary state.
+	if startSlot%params.BeaconConfig().SlotsPerEpoch == 0 {
+		startSlot += 1
 	}
 
 	filter := filters.NewFilter().SetStartSlot(startSlot).SetEndSlot(endSlot)

--- a/beacon-chain/blockchain/forkchoice/process_block.go
+++ b/beacon-chain/blockchain/forkchoice/process_block.go
@@ -382,11 +382,6 @@ func (s *Store) rmStatesBySlots(ctx context.Context, startSlot uint64, endSlot u
 	ctx, span := trace.StartSpan(ctx, "forkchoice.rmStatesBySlots")
 	defer span.End()
 
-	// Do not remove genesis state and epoch boundary state.
-	if startSlot%params.BeaconConfig().SlotsPerEpoch == 0 {
-		startSlot++
-	}
-
 	filter := filters.NewFilter().SetStartSlot(startSlot).SetEndSlot(endSlot)
 	roots, err := s.db.BlockRoots(ctx, filter)
 	if err != nil {

--- a/beacon-chain/blockchain/forkchoice/process_block_test.go
+++ b/beacon-chain/blockchain/forkchoice/process_block_test.go
@@ -334,9 +334,12 @@ func TestRemoveStateBySlots(t *testing.T) {
 		if err != nil {
 			t.Fatal(err)
 		}
-		// Also verifies genesis state didnt get deleted
-		if s != nil && s.Slot != 0 && s.Slot < endSlot {
-			t.Errorf("State with slot %d should not be in DB", s.Slot)
+		// Also verifies boundary state didnt get deleted
+		if s != nil {
+			isBoundary := s.Slot%params.BeaconConfig().SlotsPerEpoch == 0
+			if !isBoundary && s.Slot < endSlot {
+				t.Errorf("State with slot %d should not be in DB", s.Slot)
+			}
 		}
 	}
 }

--- a/beacon-chain/blockchain/forkchoice/process_block_test.go
+++ b/beacon-chain/blockchain/forkchoice/process_block_test.go
@@ -276,7 +276,7 @@ func TestStore_SavesNewBlockAttestations(t *testing.T) {
 	}
 }
 
-func TestRemoveStateBySlots(t *testing.T) {
+func TestRemoveStateSinceLastFinalized(t *testing.T) {
 	ctx := context.Background()
 	db := testDB.SetupDB(t)
 	defer testDB.TeardownDB(t, db)
@@ -309,7 +309,7 @@ func TestRemoveStateBySlots(t *testing.T) {
 	// New finalized epoch: 1
 	finalizedEpoch := uint64(1)
 	endSlot := helpers.StartSlot(finalizedEpoch+1) - 1 // Inclusive
-	if err := store.rmStatesBySlots(ctx, 0, endSlot); err != nil {
+	if err := store.rmStatesSinceLastFinalized(ctx, 0, endSlot); err != nil {
 		t.Fatal(err)
 	}
 	for _, r := range blockRoots {
@@ -326,7 +326,7 @@ func TestRemoveStateBySlots(t *testing.T) {
 	// New finalized epoch: 5
 	newFinalizedEpoch := uint64(5)
 	endSlot = helpers.StartSlot(newFinalizedEpoch+1) - 1 // Inclusive
-	if err := store.rmStatesBySlots(ctx, helpers.StartSlot(finalizedEpoch+1), endSlot); err != nil {
+	if err := store.rmStatesSinceLastFinalized(ctx, helpers.StartSlot(finalizedEpoch+1), endSlot); err != nil {
 		t.Fatal(err)
 	}
 	for _, r := range blockRoots {

--- a/beacon-chain/blockchain/forkchoice/process_block_test.go
+++ b/beacon-chain/blockchain/forkchoice/process_block_test.go
@@ -309,7 +309,7 @@ func TestRemoveStateSinceLastFinalized(t *testing.T) {
 	// New finalized epoch: 1
 	finalizedEpoch := uint64(1)
 	endSlot := helpers.StartSlot(finalizedEpoch+1) - 1 // Inclusive
-	if err := store.rmStatesSinceLastFinalized(ctx, 0, endSlot); err != nil {
+	if err := store.rmStatesOlderThanLastFinalized(ctx, 0, endSlot); err != nil {
 		t.Fatal(err)
 	}
 	for _, r := range blockRoots {
@@ -326,7 +326,7 @@ func TestRemoveStateSinceLastFinalized(t *testing.T) {
 	// New finalized epoch: 5
 	newFinalizedEpoch := uint64(5)
 	endSlot = helpers.StartSlot(newFinalizedEpoch+1) - 1 // Inclusive
-	if err := store.rmStatesSinceLastFinalized(ctx, helpers.StartSlot(finalizedEpoch+1), endSlot); err != nil {
+	if err := store.rmStatesOlderThanLastFinalized(ctx, helpers.StartSlot(finalizedEpoch+1), endSlot); err != nil {
 		t.Fatal(err)
 	}
 	for _, r := range blockRoots {

--- a/beacon-chain/db/kv/state.go
+++ b/beacon-chain/db/kv/state.go
@@ -11,6 +11,7 @@ import (
 	"github.com/gogo/protobuf/proto"
 	"github.com/pkg/errors"
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
 	"go.opencensus.io/trace"
 )
 
@@ -107,15 +108,20 @@ func (k *Store) DeleteState(ctx context.Context, blockRoot [32]byte) error {
 	defer span.End()
 
 	return k.db.Batch(func(tx *bolt.Tx) error {
-		finalizedCheckpt, err := k.FinalizedCheckpoint(ctx)
-		if err != nil {
-			return err
-		}
-
 		bkt := tx.Bucket(blocksBucket)
 		genesisBlockRoot := bkt.Get(genesisBlockRootKey)
+
+		bkt = tx.Bucket(checkpointBucket)
+		enc := bkt.Get(finalizedCheckpointKey)
+		checkpoint := &ethpb.Checkpoint{}
+		if enc == nil {
+			checkpoint = &ethpb.Checkpoint{Root: genesisBlockRoot}
+		} else {
+			proto.Unmarshal(enc, checkpoint)
+		}
+
 		// Safe guard against deleting genesis or finalized state.
-		if bytes.Equal(blockRoot[:], finalizedCheckpt.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) {
+		if bytes.Equal(blockRoot[:], checkpoint.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) {
 			return errors.New("could not delete genesis or finalized state")
 		}
 

--- a/beacon-chain/db/kv/state.go
+++ b/beacon-chain/db/kv/state.go
@@ -1,6 +1,7 @@
 package kv
 
 import (
+	"bytes"
 	"context"
 	"fmt"
 	"strings"
@@ -104,8 +105,21 @@ func (k *Store) SaveState(ctx context.Context, state *pb.BeaconState, blockRoot 
 func (k *Store) DeleteState(ctx context.Context, blockRoot [32]byte) error {
 	ctx, span := trace.StartSpan(ctx, "BeaconDB.DeleteState")
 	defer span.End()
+
+	finalizedCheckpt, err := k.FinalizedCheckpoint(ctx)
+	if err != nil {
+		return err
+	}
+
 	return k.db.Batch(func(tx *bolt.Tx) error {
-		bkt := tx.Bucket(stateBucket)
+		bkt := tx.Bucket(blocksBucket)
+		genesisBlockRoot := bkt.Get(genesisBlockRootKey)
+		// Safe guard against deleting genesis or finalized state.
+		if bytes.Equal(blockRoot[:], finalizedCheckpt.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) {
+			return nil
+		}
+
+		bkt = tx.Bucket(stateBucket)
 		return bkt.Delete(blockRoot[:])
 	})
 }

--- a/beacon-chain/db/kv/state.go
+++ b/beacon-chain/db/kv/state.go
@@ -116,7 +116,7 @@ func (k *Store) DeleteState(ctx context.Context, blockRoot [32]byte) error {
 		genesisBlockRoot := bkt.Get(genesisBlockRootKey)
 		// Safe guard against deleting genesis or finalized state.
 		if bytes.Equal(blockRoot[:], finalizedCheckpt.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) {
-			return nil
+			return errors.New("could not delete genesis or finalized state")
 		}
 
 		bkt = tx.Bucket(stateBucket)

--- a/beacon-chain/db/kv/state.go
+++ b/beacon-chain/db/kv/state.go
@@ -106,12 +106,12 @@ func (k *Store) DeleteState(ctx context.Context, blockRoot [32]byte) error {
 	ctx, span := trace.StartSpan(ctx, "BeaconDB.DeleteState")
 	defer span.End()
 
-	finalizedCheckpt, err := k.FinalizedCheckpoint(ctx)
-	if err != nil {
-		return err
-	}
-
 	return k.db.Batch(func(tx *bolt.Tx) error {
+		finalizedCheckpt, err := k.FinalizedCheckpoint(ctx)
+		if err != nil {
+			return err
+		}
+
 		bkt := tx.Bucket(blocksBucket)
 		genesisBlockRoot := bkt.Get(genesisBlockRootKey)
 		// Safe guard against deleting genesis or finalized state.

--- a/beacon-chain/db/kv/state_test.go
+++ b/beacon-chain/db/kv/state_test.go
@@ -176,15 +176,9 @@ func TestStore_DeleteGenesisState(t *testing.T) {
 	if err := db.SaveState(ctx, genesisState, genesisBlockRoot); err != nil {
 		t.Fatal(err)
 	}
-	if err := db.DeleteState(ctx, genesisBlockRoot); err != nil {
-		t.Fatal(err)
-	}
-	s, err := db.State(ctx, genesisBlockRoot)
-	if err != nil {
-		t.Fatal(err)
-	}
-	if !reflect.DeepEqual(s, genesisState) {
-		t.Error("Did not receive wanted genesis state")
+	wantedErr := "could not delete genesis or finalized state"
+	if err := db.DeleteState(ctx, genesisBlockRoot); err.Error() != wantedErr {
+		t.Error("Did not receive wanted error")
 	}
 }
 
@@ -202,14 +196,8 @@ func TestStore_DeleteFinalizedState(t *testing.T) {
 	if err := db.SaveState(ctx, finalizedState, finalizedBlockRoot); err != nil {
 		t.Fatal(err)
 	}
-	if err := db.DeleteState(ctx, finalizedBlockRoot); err != nil {
-		t.Fatal(err)
-	}
-	s, err := db.State(ctx, finalizedBlockRoot)
-	if err != nil {
-		t.Fatal(err)
-	}
-	if !reflect.DeepEqual(s, finalizedState) {
-		t.Error("Did not receive wanted finalized state")
+	wantedErr := "could not delete genesis or finalized state"
+	if err := db.DeleteState(ctx, finalizedBlockRoot); err.Error() != wantedErr {
+		t.Error("Did not receive wanted error")
 	}
 }

--- a/beacon-chain/db/kv/state_test.go
+++ b/beacon-chain/db/kv/state_test.go
@@ -162,3 +162,54 @@ func TestStore_StatesBatchDelete(t *testing.T) {
 		}
 	}
 }
+
+func TestStore_DeleteGenesisState(t *testing.T) {
+	db := setupDB(t)
+	defer teardownDB(t, db)
+	ctx := context.Background()
+
+	genesisBlockRoot := [32]byte{'A'}
+	if err := db.SaveGenesisBlockRoot(ctx, genesisBlockRoot); err != nil {
+		t.Fatal(err)
+	}
+	genesisState := &pb.BeaconState{Slot: 100}
+	if err := db.SaveState(ctx, genesisState, genesisBlockRoot); err != nil {
+		t.Fatal(err)
+	}
+	if err := db.DeleteState(ctx, genesisBlockRoot); err != nil {
+		t.Fatal(err)
+	}
+	s, err := db.State(ctx, genesisBlockRoot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(s, genesisState) {
+		t.Error("Did not receive wanted genesis state")
+	}
+}
+
+func TestStore_DeleteFinalizedState(t *testing.T) {
+	db := setupDB(t)
+	defer teardownDB(t, db)
+	ctx := context.Background()
+
+	finalizedBlockRoot := [32]byte{'A'}
+	finalizedCheckpoint := &ethpb.Checkpoint{Root: finalizedBlockRoot[:]}
+	if err := db.SaveFinalizedCheckpoint(ctx, finalizedCheckpoint); err != nil {
+		t.Fatal(err)
+	}
+	finalizedState := &pb.BeaconState{Slot: 100}
+	if err := db.SaveState(ctx, finalizedState, finalizedBlockRoot); err != nil {
+		t.Fatal(err)
+	}
+	if err := db.DeleteState(ctx, finalizedBlockRoot); err != nil {
+		t.Fatal(err)
+	}
+	s, err := db.State(ctx, finalizedBlockRoot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(s, finalizedState) {
+		t.Error("Did not receive wanted finalized state")
+	}
+}

--- a/tools/interop/export-genesis/main.go
+++ b/tools/interop/export-genesis/main.go
@@ -10,7 +10,6 @@ import (
 	"github.com/prysmaticlabs/prysm/beacon-chain/db"
 )
 
-
 // A basic tool to extract genesis.ssz from existing beaconchain.db.
 // ex:
 //   bazel run //tools/interop/export-genesis:export-genesis -- /tmp/data/beaconchaindata /tmp/genesis.ssz

